### PR TITLE
Use early stopping

### PR DESCRIPTION
**Pull request by dimidd**
* keep track of the best validation accuracy
* stop if validation accuracy doesn't improve for a pre-specified number of
epochs.

Motivation
While the training error may constantly improve with more epochs,
but it might lead to overfitting. In order to prevent this, we measure the
generalization error on the validation set, and stop the training process if
it starts to deteriorate. Early stopping can be viewed as a hyper-parameter.

Signed-off-by: Dimid Duchovny <dimidd@gmail.com>

### CODE REVIEW

**Comment by dimidd** [comment]
I just realized that there's no test set, i.e. the data is split to just two sets: train and validation (which is actually the test set). I'll re-split to three to make this work.

**Comment by dimidd** [comment]
Noah, this is now ready for review. I got slightly better results on the test set (95.4% vs 93% in the thesis). However, there are some confounding factors, e.g. I've used a newer version of dynet and its dependent libraries, so I'm not sure the results are directly comparable.

**Comment by nsantacruz** [comment]
Sorry, I was on vacation and now I'm back. This is a great improvement! I reviewed the code and it looks good
